File: client/notify.lua
--[[ ═══════════════════════════════════════════════════════════════════════════
     🐺 LXR-CORE — Unified Notifications (client)
     ═══════════════════════════════════════════════════════════════════════════
     One entry point, many callers:
       LXRCore.Functions.Notify('text')                       → tip, 4s
       LXRCore.Functions.Notify('text', 'error', 6000)
       LXRCore.Functions.Notify({ title = 'x', description = 'y', type = 'success' })
       LXRCore.Functions.Notify(9, 'text', 4000, nil, dict, icon, color)   (legacy QBR numeric id)
     Server: TriggerClientEvent('LXRCore:Notify', src, message, type, duration)
     Backend: Config.Notify.backend = 'native' (RedM feed via notify.js),
              'ox_lib' (lib.notify when started) or 'event' (re-broadcast as
              'lxr-notify:client:show' for a custom HUD).
     ═══════════════════════════════════════════════════════════════════════════
     © 2026 iBoss21 / LXRCore — All Rights Reserved
     ═══════════════════════════════════════════════════════════════════════════ ]]

local RES = LXRCore.ResourceName
local ICONS = {
    success = { dict = 'generic_textures', icon = 'tick', color = 'COLOR_GREEN' },
    error   = { dict = 'generic_textures', icon = 'cross', color = 'COLOR_RED' },
    warning = { dict = 'generic_textures', icon = 'hud_menu_4a', color = 'COLOR_YELLOW' },
    inform  = { dict = 'generic_textures', icon = 'hud_menu_4a', color = 'COLOR_WHITE' },
}

-- Legacy QBR numeric ids
local LEGACY = {
    [1] = 'ShowTooltip', [2] = 'DisplayRightText', [3] = 'ShowObjective', [4] = 'ShowBasicTopNotification',
    [5] = 'ShowSimpleCenterText', [6] = 'ShowLocationNotification', [7] = 'ShowTopNotification',
    [8] = 'ShowAdvancedLeftNotification', [9] = 'ShowAdvancedRightNotification',
}

local function loadTexture(dict)
    if not dict or dict == '' then return false end
    if not Citizen.InvokeNative(0x7332461FC59EB7EC, dict) then return false end -- DoesStreamedTextureDictExist
    RequestStreamedTextureDict(dict, true)
    local tries = 0
    while not HasStreamedTextureDictLoaded(dict) and tries < 50 do Wait(10) tries = tries + 1 end
    return HasStreamedTextureDictLoaded(dict)
end

local function native(opts)
    local kind = ICONS[opts.type] or ICONS.inform
    local title, desc = opts.title, opts.description
    if title and desc then
        loadTexture(kind.dict)
        exports[RES]:ShowAdvancedLeftNotification(title, desc, kind.dict, kind.icon, opts.duration, kind.color)
    else
        exports[RES]:ShowTooltip(title or desc or '', opts.duration)
    end
end

local function oxlib(opts)
    TriggerEvent('ox_lib:notify', { title = opts.title, description = opts.description, type = opts.type, duration = opts.duration })
end

---Normalise every accepted signature into { title, description, type, duration }.
local function normalise(a, b, c, d, e, f, g)
    local opts
    if type(a) == 'table' then
        opts = { title = a.title or a.text or a.message, description = a.description or a.subtext, type = a.type, duration = a.duration or a.length }
    elseif type(a) == 'number' and LEGACY[a] then
        -- (id, text, duration, subtext, dict, icon, color)
        return { legacy = LEGACY[a], text = b, duration = c, subtext = d, dict = e, icon = f, color = g }
    else
        opts = { title = a, type = b, duration = c }
    end
    opts.type = type(opts.type) == 'string' and opts.type:lower() or 'inform'
    if opts.type == 'info' then opts.type = 'inform' end
    opts.duration = tonumber(opts.duration) or (Config.Notify and Config.Notify.duration) or 4000
    if opts.title then opts.title = tostring(opts.title) end
    if opts.description then opts.description = tostring(opts.description) end
    return opts
end

function LXRCore.Functions.Notify(...)
    local opts = normalise(...)
    if opts.legacy then
        local fn = opts.legacy
        if fn == 'ShowAdvancedLeftNotification' then
            loadTexture(opts.dict or 'generic_textures')
            exports[RES][fn](nil, opts.text, opts.subtext or '', opts.dict or 'generic_textures', opts.icon or 'tick', opts.duration, opts.color)
        elseif fn == 'ShowAdvancedRightNotification' then
            loadTexture(opts.dict or 'generic_textures')
            exports[RES][fn](nil, opts.text, opts.dict or 'generic_textures', opts.icon or 'tick', opts.color or 'COLOR_WHITE', opts.duration)
        elseif fn == 'ShowTopNotification' or fn == 'ShowLocationNotification' then
            exports[RES][fn](nil, opts.text, opts.subtext or '', opts.duration)
        else
            exports[RES][fn](nil, opts.text, opts.duration)
        end
        return
    end
    local backend = (Config.Notify and Config.Notify.backend) or 'native'
    if backend == 'ox_lib' and GetResourceState('ox_lib') == 'started' then
        return oxlib(opts)
    elseif backend == 'event' then
        return TriggerEvent('lxr-notify:client:show', opts)
    end
    native(opts)
end

RegisterNetEvent('lxr:client:notify', function(...) LXRCore.Functions.Notify(...) end)
RegisterNetEvent('LXRCore:Notify', function(...) LXRCore.Functions.Notify(...) end) -- legacy name

-- RSG resources call lib.notify; when ox_lib is absent, route its event to us.
if GetResourceState('ox_lib') ~= 'started' then
    RegisterNetEvent('ox_lib:notify', function(data)
        if type(data) == 'table' then native(normalise(data)) end
    end)
end

exports('Notify', LXRCore.Functions.Notify)

File: client/notify.js
/*  ═══════════════════════════════════════════════════════════════════════════
    🐺 LXR-CORE — Native RedM UI Feed Notifications (client, JavaScript)
    ═══════════════════════════════════════════════════════════════════════════
    The RDR3 feed natives take pointer structs, which Lua cannot build without a
    DataView shim; JavaScript has one natively, so this file owns the struct
    marshalling and exposes plain exports consumed by client/notify.lua:

      ShowTooltip(text, duration)                      top-left tip
      DisplayRightText(text, duration)                 right-hand tip
      ShowObjective(text, duration)                    bottom objective
      ShowBasicTopNotification(text, duration)         top banner
      ShowSimpleCenterText(text, duration)             center text
      ShowTopNotification(title, subtitle, duration)   top banner with subtitle
      ShowLocationNotification(text, location, duration)
      ShowAdvancedLeftNotification(title, subtitle, dict, icon, duration)
      ShowAdvancedRightNotification(text, dict, icon, color, duration)

    Struct layouts follow the public RDR3 native documentation.
    ═══════════════════════════════════════════════════════════════════════════
    © 2026 iBoss21 / LXRCore — All Rights Reserved
    ═══════════════════════════════════════════════════════════════════════════ */

(() => {
  'use strict';

  const str = (v) => CreateVarString(10, 'LITERAL_STRING', String(v ?? ''));
  const hash = (v) => BigInt(GetHashKey(String(v ?? '')));
  const ms = (v, d = 4000) => { const n = Number(v); return Number.isFinite(n) && n > 0 ? Math.floor(n) : d; };

  const duration = (d) => {
    const s = new DataView(new ArrayBuffer(48));
    s.setInt32(0, ms(d), true);
    return s;
  };

  const body = (size, writer) => {
    const s = new DataView(new ArrayBuffer(size));
    writer(s);
    return s;
  };

  // 0x049D5C615BD38BAD  UiFeedPostSampleNshTooltip
  exports('ShowTooltip', (text, d) => {
    const b = body(16, (s) => s.setBigInt64(8, BigInt(str(text)), true));
    Citizen.invokeNative('0x049D5C615BD38BAD', duration(d), b, 1);
  });

  // 0xB2920B9760F0F36B  UiFeedPostSampleNshRightText
  exports('DisplayRightText', (text, d) => {
    const b = body(16, (s) => s.setBigInt64(8, BigInt(str(text)), true));
    Citizen.invokeNative('0xB2920B9760F0F36B', duration(d), b, 1);
  });

  // 0xCEDBF17EFCC0E4A4  UiFeedPostObjective
  exports('ShowObjective', (text, d) => {
    const b = body(16, (s) => s.setBigInt64(8, BigInt(str(text)), true));
    Citizen.invokeNative('0xCEDBF17EFCC0E4A4', duration(d), b, 1);
  });

  // 0x860DDFE97CC94DF0  UiFeedPostOneTextShard
  exports('ShowBasicTopNotification', (text, d) => {
    const b = body(48, (s) => s.setBigInt64(8, BigInt(str(text)), true));
    Citizen.invokeNative('0x860DDFE97CC94DF0', duration(d), b, 1);
  });

  // 0x893128CDB4B81FBB  UiFeedPostSimpleText (center)
  exports('ShowSimpleCenterText', (text, d, color) => {
    const b = body(24, (s) => {
      s.setBigInt64(8, BigInt(str(text)), true);
      s.setBigInt64(16, hash(color || 'COLOR_PURE_WHITE'), true);
    });
    Citizen.invokeNative('0x893128CDB4B81FBB', duration(d), b, 1);
  });

  // 0xA6F4216AB10EB08E  UiFeedPostTwoTextShard
  exports('ShowTopNotification', (title, subtitle, d) => {
    const b = body(48, (s) => {
      s.setBigInt64(8, BigInt(str(title)), true);
      s.setBigInt64(16, BigInt(str(subtitle)), true);
    });
    Citizen.invokeNative('0xA6F4216AB10EB08E', duration(d), b, 1, 1);
  });

  // 0xD05590C1AB38F068  UiFeedPostSampleNshLocation
  exports('ShowLocationNotification', (text, location, d) => {
    const b = body(24, (s) => {
      s.setBigInt64(8, BigInt(str(location)), true);
      s.setBigInt64(16, BigInt(str(text)), true);
    });
    Citizen.invokeNative('0xD05590C1AB38F068', duration(d), b, 1, 1);
  });

  // 0x26E87218390E6729  UiFeedPostSampleNshMessage (left, with texture)
  exports('ShowAdvancedLeftNotification', (title, subtitle, dict, icon, d, color) => {
    const b = body(56, (s) => {
      s.setBigInt64(8, BigInt(str(title)), true);
      s.setBigInt64(16, BigInt(str(subtitle)), true);
      s.setBigInt64(32, hash(dict || 'generic_textures'), true);
      s.setBigInt64(40, hash(icon || 'tick'), true);
      s.setBigInt64(48, hash(color || 'COLOR_WHITE'), true);
    });
    Citizen.invokeNative('0x26E87218390E6729', duration(d), b, 1, 1);
  });

  // 0xB249EBCB30DD88E0  UiFeedPostSampleNshRight (with texture + sound)
  exports('ShowAdvancedRightNotification', (text, dict, icon, color, d) => {
    const s1 = duration(d);
    s1.setBigInt64(8, BigInt(str('Transaction_Feed_Sounds')), true);
    s1.setBigInt64(16, BigInt(str('Transaction_Positive')), true);
    const b = body(80, (s) => {
      s.setBigInt64(8, BigInt(str(text)), true);
      s.setBigInt64(16, BigInt(str(dict || 'generic_textures')), true);
      s.setBigInt64(24, hash(icon || 'tick'), true);
      s.setBigInt64(40, hash(color || 'COLOR_WHITE'), true);
      s.setInt32(48, 0, true);
    });
    Citizen.invokeNative('0xB249EBCB30DD88E0', s1, b, 1);
  });
})();
